# Neural Networks

In [1]:
import pandas as pd

In [2]:
y_train = pd.read_feather('data/processed/y_train.feather')['voted_up'].to_numpy()
X_train = pd.read_feather(f'data/processed/X_train_bigram.feather').to_numpy()

In [3]:
y_test = pd.read_feather('data/processed/y_test.feather')['voted_up'].to_numpy()
X_test = pd.read_feather(f'data/processed/X_test_bigram.feather').to_numpy()

## Baseline Neural Network

In [5]:
model = Sequential()

# hidden layers
model.add(LSTM(500, input_dim=8000, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(100, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(50, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(100, activation='relu'))

# output layer
model.add(Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [ ]:
model.fit(X_train, y_train, epochs=3, batch_size=32, validation_data=(X_test, y_test))
model.evaluate(X_test, y_test)

In [4]:
from sklearn.metrics import accuracy_score, precision_score, recall_score
from tensorflow.keras.layers import Dense, Dropout, LSTM
from tensorflow.keras import Sequential

In [8]:
model = Sequential()

# hidden layers
model.add(Dense(500, input_dim=8000, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(100, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(50, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(100, activation='relu'))

# output layer
model.add(Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [9]:
model.fit(X_train, y_train, epochs=3, batch_size=32, validation_data=(X_test, y_test))
model.evaluate(X_test, y_test)

Epoch 1/3
1828/1828 [==============================] - 61s 32ms/step - loss: 0.3386 - accuracy: 0.8589 - val_loss: 0.2210 - val_accuracy: 0.9097
Epoch 2/3
1828/1828 [==============================] - 53s 29ms/step - loss: 0.1787 - accuracy: 0.9328 - val_loss: 0.2159 - val_accuracy: 0.9190
Epoch 3/3
457/457 [==============================] - 4s 9ms/step - loss: 0.2229 - accuracy: 0.9180


[0.2228986769914627, 0.9179890751838684]

0.9179890751838684

Top performing base model: 0.914090 test accuracy

## Hyperparameter Tuning

In [5]:
from talos import Scan

In [6]:
def dense_network(x_train, y_train, x_val, y_val, params):
    model = Sequential()

    # hidden layers
    model.add(Dense(params['dense'], input_dim=8000, activation=params['activation1']))
    model.add(Dropout(params['dropout']))
    model.add(Dense(params['dense']*2, activation=params['activation1']))
    model.add(Dropout(params['dropout']))
    model.add(Dense(params['dense']*0.5, activation=params['activation1']))
    model.add(Dropout(params['dropout']))
    model.add(Dense(params['dense']*0.75, activation=params['activation1']))
    # output layer
    model.add(Dense(1, activation=params['activation2']))

    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    
    out = model.fit(x_train, y_train,
                    validation_data=(x_val, y_val),
                    epochs=5,
                    verbose=0)

    return out, model

In [7]:
params = {'dropout': [0.25, 0.5, 0.75], 
          'dense': [10, 50, 100, 500], 
          'activation1': ['relu', 'elu'], 
          'activation2': ['sigmoid', 'tanh']}

In [8]:
results = Scan(X_train, y_train, params=params, model=dense_network, experiment_name='grid')
results.best_model(metric='accuracy')

100%|██████████| 48/48 [1:14:46<00:00, 93.47s/it] 


In [9]:
pd.read_csv('grid/022421165029.csv').sort_values('val_accuracy', ascending=False)

round_epochs       loss  accuracy   val_loss  val_accuracy activation1  \
10             5   0.038711  0.988078   0.414048      0.911247        relu   
22             5   0.107399  0.980114   0.465812      0.910050        relu   
41             5   0.338274  0.902475   0.373536      0.909252         elu   
11             5   0.139434  0.951653   0.270693      0.908853        relu   
4              5   0.113932  0.960887   0.254306      0.908682        relu   
23             5   0.192981  0.952728   0.379954      0.907712        relu   
9              5   0.018463  0.994503   0.530807      0.907655        relu   
44             5   0.345217  0.908875   0.448334      0.907484         elu   
18             5   0.041759  0.990765   0.592345      0.907085        relu   
26             5   0.318125  0.877287   0.241634      0.906743         elu   
29             5   0.195315  0.930374   0.244876      0.906401         elu   
36             5   0.235776  0.924266   0.328179      0.905489         elu   
40             5   0.257518  0.929348   0.385643      0.905261         elu   
15             5   0.051648  0.987247   0.564256      0.904976        relu   
31             5   0.149231  0.945179   0.267483      0.904064         elu   
25             5   0.193293  0.933110   0.256739      0.903950         elu   
35             5   0.166995  0.936408   0.256267      0.903722         elu   
43             5   0.291239  0.923265   0.467894      0.902924         elu   
32             5   0.180114  0.935651   0.258423      0.902867         elu   
7              5   0.081260  0.973347   0.306307      0.902810        relu   
42             5   0.217653  0.934991   0.435936      0.902525         elu   
27             5   0.126879  0.949723   0.316994      0.901955         elu   
24             5   0.148680  0.951384   0.275825      0.901784         elu   
33             5   0.114917  0.951017   0.323617      0.901442         elu   
30             5   0.115778  0.953265   0.324620      0.901214         elu   
6              5   0.029110  0.990717   0.432521      0.901043        relu   
13             5   0.282703  0.862043   0.280506      0.900815        relu   
28             5   0.156191  0.944568   0.266543      0.900758         elu   
38             5   0.440055  0.853688   0.315833      0.898364         elu   
16             5   0.106166  0.964454   0.261495      0.896996        relu   
39             5   0.242899  0.931205   0.323872      0.896825         elu   
19             5   0.081372  0.976547   0.461106      0.896426        relu   
0              5   0.149359  0.949332   0.330978      0.895514        relu   
34             5   0.131439  0.945032   0.354237      0.895001         elu   
3              5   0.065606  0.980896   0.345694      0.894260        relu   
21             5   0.038714  0.993868   1.097150      0.891239        relu   
47             5   0.382284  0.909877   0.369443      0.888617         elu   
1              5   0.245896  0.904160   0.269550      0.885994        relu   
37             5   0.327380  0.881489   0.288364      0.876817         elu   
5              5   0.232989  0.913126   0.625178      0.872428        relu   
46             5   2.087970  0.863142   2.587984      0.830303         elu   
45             5   2.373882  0.843989   2.827911      0.814627         elu   
14             5   0.451831  0.806440   0.460589      0.804651        relu   
2              5   0.432529  0.806440   0.364455      0.804651        relu   
8              5   0.202846  0.926881   0.574749      0.655988        relu   
20             5   0.178975  0.933745   1.036396      0.256228        relu   
17             5   0.212452  0.920309   0.953912      0.215072        relu   
12             5  12.439324  0.193560  12.411721      0.195349        relu   

   activation2  dense  dropout  
10     sigmoid    500     0.50  
22        tanh    500     0.50  
41        tanh     50     0.75  
11     sigmoid    500     0.75  
4      sigmoid  

## Final Model

In [5]:
model = Sequential()

# hidden layers
model.add(Dense(500, input_dim=8000, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(250, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(125, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(250, activation='relu'))

# output layer
model.add(Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [6]:
model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.1)

Epoch 1/10
1645/1645 [==============================] - 53s 31ms/step - loss: 0.3298 - accuracy: 0.8628 - val_loss: 0.2276 - val_accuracy: 0.9082
Epoch 2/10
1645/1645 [==============================] - 49s 30ms/step - loss: 0.1735 - accuracy: 0.9360 - val_loss: 0.2266 - val_accuracy: 0.9099
Epoch 3/10
1645/1645 [==============================] - 49s 30ms/step - loss: 0.1182 - accuracy: 0.9577 - val_loss: 0.2352 - val_accuracy: 0.9082
Epoch 4/10
1645/1645 [==============================] - 49s 30ms/step - loss: 0.0719 - accuracy: 0.9752 - val_loss: 0.2849 - val_accuracy: 0.9060
Epoch 5/10
1645/1645 [==============================] - 50s 30ms/step - loss: 0.0404 - accuracy: 0.9877 - val_loss: 0.3743 - val_accuracy: 0.90660s - loss: 0.0404 - ac
Epoch 6/10
1645/1645 [==============================] - 51s 31ms/step - loss: 0.0297 - accuracy: 0.9902 - val_loss: 0.3590 - val_accuracy: 0.9061
Epoch 7/10
1645/1645 [==============================] - 49s 30ms/step - loss: 0.0236 - accuracy: 0.992

In [7]:
model.evaluate(X_test, y_test)

457/457 [==============================] - 4s 9ms/step - loss: 0.4588 - accuracy: 0.9112


[0.4588494598865509, 0.9112175107002258]

In [8]:
model.save('final_model/model.h5')
model.save_weights('final_model/model_weights.h5')